### PR TITLE
Code cleanup on greeting processing

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Code cleanup on  greeting procssing

Before
----------------------------------------
Each of the three greetings parsed separately, with separate calls to the greeting processor

After
----------------------------------------
one combined call

Technical Details
----------------------------------------
Ideally `sort_name` & `display_name` would be added in - but let's see what happens...

Comments
----------------------------------------
 I thought this would speed up performance but in testing so far it doesn't - however, the processing of greetings has a significant speed impact so I'm investigating why - which doesn't negate this PR as a code cleanup / move away from legacy functions

### CODE REVIEW

**Comment by seamuslee001** [comment]
@johntwyman @andrew-cormick-dockery @anthonyblond you folks might be interested in this

**Comment by eileenmcnaughton** [comment]
@seamuslee001 the test fails are hard to know what to do with - I removed an extraneous db lookup that is no longer required for the internal workings of the `BAO:create` - but some tests check the output from them. I suspect this is just lazy tests - ie not doing a `get` but haven't completely decided

**Comment by eileenmcnaughton** [comment]
Test Result (5 failures / +5)api\v4\Action\UpdateContactTest.testUpdateWithIdInWhereapi\v4\Action\UpdateContactTest.testUpdateWithIdInValuesapi_v3_ContactTest.testCreateApiKey with data set "APIv3"api_v3_ContactTest.testCreateApiKey with data set "APIv4"CRM_Core_BAO_ActionScheduleTest.testContactModifiedAnniversary
--


[Test Result](https://test.civicrm.org/job/CiviCRM-Core-PR/50492/testReport/) (5 failures / +5)
[api\v4\Action\UpdateContactTest.testUpdateWithIdInWhere](https://test.civicrm.org/job/CiviCRM-Core-PR/50492/testReport/junit/(root)/api_v4_Action_UpdateContactTest/testUpdateWithIdInWhere/)
[api\v4\Action\UpdateContactTest.testUpdateWithIdInValues](https://test.civicrm.org/job/CiviCRM-Core-PR/50492/testReport/junit/(root)/api_v4_Action_UpdateContactTest/testUpdateWithIdInValues/)
[api_v3_ContactTest.testCreateApiKey with data set "APIv3"](https://test.civicrm.org/job/CiviCRM-Core-PR/50492/testReport/junit/(root)/api_v3_ContactTest/testCreateApiKey_with_data_set__APIv3_/)
[api_v3_ContactTest.testCreateApiKey with data set "APIv4"](https://test.civicrm.org/job/CiviCRM-Core-PR/50492/testReport/junit/(root)/api_v3_ContactTest/testCreateApiKey_with_data_set__APIv4_/)
[CRM_Core_BAO_ActionScheduleTest.testContactModifiedAnniversary](https://test.civicrm.org/job/CiviCRM-Core-PR/50492/testReport/junit/(root)/CRM_Core_BAO_ActionScheduleTest/testContactModifiedAnniversary/)

**Comment by eileenmcnaughton** [comment]
Note that pre-existing test cover is solid here. 

My performance testing has some queries eliminated here (perhtaps fifteen - [this query output](https://docs.google.com/spreadsheets/d/1vKBDVlY38tp_zzoRo0w1OYA-edsGWwGWFutqbHB0O90/edit#gid=0) shows the queries before patch three, which eliminated the beige one) 

However, the performance problem is not notably improved & I'm going to do some php profiling since we are seeing a contact create take about three times as long (150 ms vs 45ms) and the queries only seem to add up to a couple of ms. My guess is that there might be something in there that is php-processing instensive and does not have caching. That is non-blocking on this. I am happy this is an improvement

**Comment by eileenmcnaughton** [comment]
This latest is giving me a significant improvement. My import test set add 500 contributions & updates 500 contacts. My best speed is 667 per minute - with `skip_greeting_processing` it is 234 per minute. With this patch I get to 526 per minute

The speed difference appears to be in the php layer, not the extra queries - with apiv4 metadata functions being a possible culprit - there is something weird in that 5 api calls on the same contact seems not much different to 1,  - but one being markedly different to none. That would not be so odd except that it is on a per contact basis - perhaps some cache-clear per contact is an issue, but note that I have opportunistic cache flushing off...

**Comment by eileenmcnaughton** [comment]
ohh fun - I just got it back to 667 per minute... with a cache fix... pushing up now

**Comment by eileenmcnaughton** [comment]
oh wow - the cache fix with NO OTHER CHANGES gets me from 234 per minute to 577 per minute....

**Comment by eileenmcnaughton** [comment]
Fixed another cache miss - 612 per minute

Note that I'm using Redis caching - for sites using Array caching there is likely to be no change in these caching fixes as the difference is use of the array cache facade

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
I've moved all the other commits out of this PR & this PR is now primarily a refactor / code cleanup of legacy functions.

It WILL remove some db queries but these proved to be the least compelling of the performance improvements I trialled.

Note test cover is super solid on this - including in the tests 'touched' in this PR (I only removed incorrect throws tags)

File: CRM/Contact/BAO/Contact.php [hunk 1]
```diff
@@ -2725,64 +2727,48 @@ public static function updateGreetingsOnTokenFieldChange($updatedFields, $contac
...
-        $contact->id,
-        'CRM_Contact_BAO_Contact'
-      );
-      $addresseeString = CRM_Core_DAO::escapeString(CRM_Utils_String::stripSpaces($addresseeString));
-      $updateQueryString[] = " addressee_display = '$addresseeString'";
+    $tokenProcessor->evaluate();
+    $row = $tokenProcessor->getRow(0);
+    foreach ($greetings as $greetingKey => $greetingString) {
+      $parsedGreeting = CRM_Core_DAO::escapeString(CRM_Utils_String::stripSpaces($row->render($greetingKey)));
+      $updateQueryString[] = " $greetingKey = '$parsedGreeting'";
```
Comment (seamuslee001) [inline CRM/Contact/BAO/Contact.php:2766]: @eileenmcnaughton the one thing I wonder bout here is should we do a test for checking if $parsedGreeting != current passed greeting?

Comment (eileenmcnaughton) [reply CRM/Contact/BAO/Contact.php:2766]: hmm yeah - could do - I think it is no change at the moment - ie the check is not there already - so I'd probably look at it as a follow up

**Comment by seamuslee001** [comment]
ok test coverage should be fine here

**Pull request merged**